File: IOS/DTOs/OptionSet/MaestriPermissionSet.swift
import Foundation

public struct MaestriPermissionSet: OptionSet, Codable, Hashable, Sendable {
    
    public static let none: MaestriPermissionSet = []
    public static let service = MaestriPermissionSet(rawValue: 1 << 0) // разрешаем создавать свои сервайсы
    
    public static let full: MaestriPermissionSet = [.service]
    
    public var rawValue: Int
    
    public init(rawValue: Int) {
        self.rawValue = rawValue
    }
}

File: android-legacy/shared/src/main/kotlin/shared/dto/optionSet/MaestriPermissionSet.kt
package shared.dto.optionSet

import kotlinx.serialization.Serializable
import shared.serializers.MaestriPermissionSetSerializer

@Serializable(MaestriPermissionSetSerializer::class)
@JvmInline
value class MaestriPermissionSet(private val rawValue: Int) {
    companion object {
        val none: MaestriPermissionSet = MaestriPermissionSet(0)
        val service = MaestriPermissionSet(1 shl 0)

        val full: MaestriPermissionSet = MaestriPermissionSet(
            service.rawValue
        )
    }

    operator fun contains(permission: MaestriPermissionSet): Boolean {
        return rawValue and permission.rawValue != 0
    }

    operator fun plus(permission: MaestriPermissionSet): MaestriPermissionSet {
        return MaestriPermissionSet(rawValue or permission.rawValue)
    }

    operator fun minus(permission: MaestriPermissionSet): MaestriPermissionSet {
        return MaestriPermissionSet(rawValue and permission.rawValue.inv())
    }
}